Added RFC for Compute Shader support

galibzon opened the pull request:
Added RFC for Compute Shader support

galibzon committed:
Added RFC for Compute Shader support
in Material-related files.

Signed-off-by: galibzon <66021303+galibzon@users.noreply.github.com>

kh-huawei commented on rfcs/MaterialComputeShader/RFC_MaterialWithComputeShaders.md:
```diff
@@ -0,0 +1,322 @@
+    // but this is a compute shader. With the SAMPLE_LEVEL_LOD macro, the shader code
+    // will call Texture2D.SampleLevel() instead of Texture2D.Sample()
+    "Definitions": [
+        "SAMPLE_LEVEL_LOD=0"
```
To re-iterate the comment from @msat-huawei from the related PR: Hardcoding the sample level seems limiting. It would be better to use a define without value to switch the behavior, and keep the level selection separate.

I'd prefer to use SampleGrad() with hardcoded ddx/ddy=0, to get the same behavior but obvious start points for future changes, but a separate define with a default value if its not defined would work as well.

galibzon replied:
I agree, at the moment seems limiting, but it is just a starting point. Some may prefer SampleLevel(), while other would rather use SampleGrad() and there's nothing to forbid us from adding more macros and let the developer pick what they need.

Very soon in our internal project we'll have to figure out how to work with LOD and we'll provide a new PR for a more flexible to pick the LOD level at runtime or Gradients.

This is not the end of the road, it's just the beginning.

kh-huawei replied:
Fair. 
But, and this somewhat of a nitpick, to keep the impact of future PRs even lower, i would prefer a simple 'CS_SAMPLERS' define now (or so. Names are hard), and hardcode the LOD to 0 in hlsl code, with the implicit understanding that the sampling itself will change, but not the fact that it's sampling in a Compute Shader.

Changing the behavior of an existing define will be a greater impact in the future than adding new ones, especially if other devs already started building on your defines...

galibzon replied:
Sounds good. Will update the PR accordingly. https://github.com/o3de/o3de/pull/18509

kh-huawei commented on rfcs/MaterialComputeShader/RFC_MaterialWithComputeShaders.md:
```diff
@@ -0,0 +1,322 @@
+[numthreads(8,8,1)]
+void MainCS(uint3 dispatch_id: SV_DispatchThreadID)
+{
+    VsOutput IN = CalculateVsOutputFromStreamBuffers(dispatch_id.xy);
```
The Raytracing-Implementation already has to do something similar to get from a Hit to the actual geometry. but the RayTracingSceneSrg provides access to the geometry of all meshes.
Would it be possible to use that functionality, but moved to the MeshFeatureProcessor proper and adapted so it works for compute shaders?
Incidentially, this would also make a deferred pipeline more feasible..,

galibzon replied:
Yes, it's possible. That's why in the PR I added the classes `ShaderStreamBufferViewsBuilder` and `ShaderStreamBufferViewsBuilderInterface`, which do basically the same labor that the RayTracing code does to Bind Vertex Streams. In the PR I did not change the RayTracing code to use the `ShaderStreamBufferViewsBuilder`  to keep the code review to a minimum.

kh-huawei commented on rfcs/MaterialComputeShader/RFC_MaterialWithComputeShaders.md:
```diff
@@ -0,0 +1,322 @@
+        using DispatchArgumentsSetupCB = AZStd::function<void(uint32_t /*lodIndex*/, uint32_t /*meshIndex*/, uint32_t /*drawItemIdx*/, const RHI::DrawItem*, RHI::DispatchDirect&)>;
+        //! DisptachItems will be created for the DrawItems that match both the @drawListTagsFilter and @materialPipelineFilter.
+        //! Also, only DrawItems whose PipelineState is of Compute type will be considered.
+        virtual DispatchDrawItemList BuildDispatchDrawItemList(
```
How would this work with indirect dispatch? _Could_ this work with indirect dispatch in a similar manner, or would this need a separate interface somewhere?

kh-huawei commented on rfcs/MaterialComputeShader/RFC_MaterialWithComputeShaders.md:
```diff
@@ -0,0 +1,322 @@
+
+# What are the advantages of the feature?
+Developers will have the option of adding Compute shaders to MaterialTypes and MaterialPipelines.
+This is very important for alternative rendering methods like Texture Space Shading and having fine control
```
If i understand this correctly, you are creating a compute shader invocation for each (visible) object, but each shader is limited to exactly that object.
I'm not entirely sure if this going to be that useful on its own. At least not for Texture Space Shading, since there you usually need something like a V-Buffer pass, and once you have that, "rendering" each object separately again with a compute shader seems counter-intuitive.

galibzon replied:
Correct, the case of Texture Space Shading requires custom pipelines for the V-Buffer pass, etc. It is a very unique case, and in this case the Compute Shader runs for each (visible) object and yes, each shader is limited to exactly that object (to be more precise limited to that DrawItem).

It is very useful on its own (We are using it in our internal project), but requires other things like custom V-Buffer pass, etc... BUT the benefit of being able to integrate Compute Shaders with the material system is that we can leverage the material system shader code for lighting, etc.

kh-huawei reviewed (changes requested):
I generally support this RFC: having the option to use the Material-Shaders in a compute shader makes everything more flexible, and if we get nothing else, this is a good start.

But for this to be really useful, it would need to be a bit more flexible with the mip-level selection, and at least somehow support indirect dispatch.

And most features that would use this also need access to the geometry of multiple objects, but for that we also need RFC https://github.com/o3de/sig-graphics-audio/issues/152.

galibzon committed:
Per feedback from @kh-huawei, changed
the SAMPLE_LEVEL_LOD macro for CS_SAMPLERS.

Adjusted the RFC accordingly, and added remark that in the future
we can provide more flexibility on whether using SampleLevel() or
SampleGrad().

Signed-off-by: galibzon <66021303+galibzon@users.noreply.github.com>

moudgils reviewed (approved):
Supportive of this RFC. Thanks for the contribution. Looking forward to seeing the texture space pipeline in future :)
